xayn-ai/src/ranker/context.rs

```diff
@@ -1,119 +1,231 @@
-use crate::{data::document_data::CoiComponent, ranker::DocumentId};
-
-/// The context used to calculate a document's score.
-/// <https://xainag.atlassian.net/wiki/spaces/M2D/pages/770670607/Production+AI+Workflow#3.2-Context-calculations>.
-/// outlines the calculation of positive and negative distance factor.
-pub(crate) struct Context {
-    /// Average positive distance.
-    pos_avg: f32,
-    /// Maximum negative distance.
-    neg_max: f32,
+use std::{
+    collections::HashMap,
+    time::{Duration, SystemTime},
+};
+
+use displaydoc::Display;
+use thiserror::Error;
+
+use crate::{
+    coi::{
+        compute_coi_decay_factor,
+        find_closest_coi,
+        point::{CoiPoint, UserInterests},
+        RelevanceMap,
+    },
+    embedding::utils::Embedding,
+    ranker::{document::Document, Configuration},
+    utils::system_time_now,
+    CoiId,
+    DocumentId,
+};
+
+#[derive(Error, Debug, Display)]
+#[allow(clippy::enum_variant_names)]
+pub(crate) enum Error {
+    /// Not enough positive cois (expected {expected:?}, has {has:?})
+    NotEnoughPositiveCois { expected: u32, has: u32 },
+    /// Not enough negative cois (expected {expected:?}, has {has:?})
+    NotEnoughNegativeCois { expected: u32, has: u32 },
+    /// Failed to find the closest cois
+    FailedToFindTheClosestCois,
 }
 
-impl Context {
-    pub(crate) fn from_cois(cois: &[(DocumentId, CoiComponent)]) -> Self {
-        let cois_len = cois.len() as f32;
-        let pos_avg = cois.iter().map(|(_, coi)| coi.pos_distance).sum::<f32>() / cois_len;
-        let neg_max = cois
-            .iter()
-            .map(|(_, coi)| coi.neg_distance)
-            .fold(f32::MIN, f32::max); // NOTE f32::max considers NaN as smallest value
+/// Helper struct for [`find_closest_cois`].
+struct ClosestCois {
+    /// The ID of the positive centre of interest
+    pos_id: CoiId,
+    /// Distance from the positive centre of interest
+    pos_distance: f32,
+    pos_last_view: SystemTime,
+
+    /// Distance from the negative centre of interest
+    neg_distance: f32,
+    neg_last_view: SystemTime,
+}
 
-        Self { pos_avg, neg_max }
+fn find_closest_cois(
+    embedding: &Embedding,
+    user_interests: &UserInterests,
+    neighbors: usize,
+) -> Option<ClosestCois> {
+    let (pos_coi, pos_distance) = find_closest_coi(&user_interests.positive, embedding, neighbors)?;
+    let (neg_coi, neg_distance) = find_closest_coi(&user_interests.negative, embedding, neighbors)?;
```

Comment:
the `find_closest_coi()` computation is different to the pink code, i assume this will be changed later when the l2 distances are replaced with the cosine similarities?

Reply:
Yes, later we will use cosine similarities but other than that, did you find anything else that is different in the calculation?

Reply:
no, the computation looks the same (of course except for places where we will replace the l2 distance with cosine similarity). we have to double check the [order of sorting](https://github.com/xaynetwork/xayn_ai/blob/TY-2240-weighted-ranking/xayn-ai/src/ranker/mod.rs#L143-L148) once the change is done.